File: second course/PDOO/civitasRuby/pkg/civitasRuby-0.0.1/lib/sorpresa.rb
# encoding:utf-8

require_relative 'tipo_sorpresa'
require_relative 'diario'
require_relative 'mazo_sorpresas'
require_relative 'tablero'
require_relative 'jugador'
require_relative 'casilla'

module Civitas
	
	class Sorpresa # TESTED
		
		private_class_method :new
		
		def initialize(tipo, tablero, valor, texto, mazo)
			@texto = texto
			@valor = valor
			@mazo = mazo
			@tipo = tipo
			@tablero = tablero
		end
		
		def self.new_por_tablero(tipo, tablero)
			new(tipo, tablero, -1, nil, nil)
		end
		
		def self.new_por_tablero_valor(tipo, tablero, valor, texto)
			new(tipo, tablero, valor, texto, nil)
		end
		
		def self.new_por_valor(tipo, valor, texto)
			new(tipo, nil, valor, texto, nil)
		end
		
		def self.new_por_mazo(tipo, mazo)
			new(tipo, nil, -1, nil, mazo)
		end
		
		private
		
		def informe(actual, todos)
			Diario.instance.ocurre_evento(todos[actual].nombre() + " ha activado una sorpresa. ")
		end
		
		def aplicar_a_jugador_ir_carcel(actual, todos)
			if(jugador_correcto(actual, todos))
				informe(actual, todos)
				todos[actual].encarcelar(@tablero.num_casilla_carcel())
			end
		end
		
		def aplicar_a_jugador_ir_a_casilla(actual, todos)
			if(jugador_correcto(actual, todos))
				informe(actual, todos)
				
				casilla_actual = todos[actual].num_casilla_actual()
				tirada = @tablero.calcular_tirada(casilla_actual, @valor)
				nueva_pos = @tablero.nueva_posicion(casilla_actual, tirada)
				
				todos[actual].mover_a_casilla(nueva_pos)
				@tablero.casilla(nueva_pos).recibe_jugador(actual, todos)
			end
		end
		
		def aplicar_a_jugador_por_casa_hotel(actual, todos)
			if(jugador_correcto(actual, todos))
				informe(actual, todos)
				todos[actual].modificar_saldo(@valor * todos[actual].cantidad_casas_hoteles())
			end
		end
		
		public
		def aplicar_a_jugador_pagar_cobrar(actual, todos)
			if(jugador_correcto(actual, todos))
				informe(actual, todos)
				todos[actual].modificar_saldo(@valor)
			end
		end
		
		private
		def aplicar_a_jugador_por_jugador(actual, todos)
			if(jugador_correcto(actual, todos))
				informe(actual, todos) 
				
				pagar = Sorpresa.new_por_valor(Tipo_sorpresa::PAGARCOBRAR, -1 * @valor, "PAGAR")
				cobrar = Sorpresa.new_por_valor(Tipo_sorpresa::PAGARCOBRAR, (todos.size() - 1) * @valor, "COBRAR")
				
				for i in (0 ... todos.size())
					if(todos[i] != todos[actual])
						pagar.aplicar_a_jugador_pagar_cobrar(i, todos)
					end
				end
				
				cobrar.aplicar_a_jugador_pagar_cobrar(actual, todos)
				
			end
		end
		
		def aplicar_a_jugador_salir_carcel(actual, todos)
			if(jugador_correcto(actual, todos))
				informe(actual, todos)
				
				hay_salvo_conducto = false
				
				for i in (0 ... todos.size())
					if(todos[i].tiene_salvoconducto())
						hay_salvo_conducto = true
					end
				end
				
				if(!hay_salvo_conducto)
					todos[actual].obtener_salvoconducto(self)
					salir_del_mazo()
				end
				
			end
		end
		
		
		public
		
		def jugador_correcto(actual, todos)
			return actual >= 0 && actual < todos.size()
		end
		
		def aplicar_a_jugador(actual, todos)
			case @tipo
				when Tipo_sorpresa::IRCARCEL
          aplicar_a_jugador_ir_carcel(actual,todos)
				when Tipo_sorpresa::IRCASILLA
         aplicar_a_jugador_ir_a_casilla(actual, todos)
				when Tipo_sorpresa::SALIRCARCEL
          aplicar_a_jugador_salir_carcel(actual, todos)
				when Tipo_sorpresa::PAGARCOBRAR
          aplicar_a_jugador_pagar_cobrar(actual, todos)
				when Tipo_sorpresa::PORCASAHOTEL
          aplicar_a_jugador_por_casa_hotel(actual, todos)
				when Tipo_sorpresa::PORJUGADOR               
          aplicar_a_jugador_por_jugador(actual,todos)
      end
		end
		
		def salir_del_mazo()
			
			if(@tipo == Tipo_sorpresa::SALIRCARCEL)
				@mazo.inhabilitar_carta_especial(self)
			end
		end
		
		def to_string()
			return "Sorpresa: #{@tipo} 
							texto: #{@texto}\n"
		end
		
		def usada()
			if((@tipo == Tipo_sorpresa::SALIRCARCEL))
				@mazo.habilitar_carta_especial(self)
			end
		end
		
		
	end
end

File: second course/PDOO/civitasJava/src/civitas/Sorpresa.java
package civitas;

import java.util.ArrayList;

public class Sorpresa{
  
  private String texto;
  private int valor;
  
  //Atributos de referencia
  private MazoSorpresas mazo;
  private TipoSorpresa tipo;
  private Tablero tablero;
  
  
  private void init(){
		mazo = null;
		tablero = null;
		valor = -1;
  }
  
  Sorpresa(TipoSorpresa tipo, Tablero tablero, int valor, String texto){ //Envia a otra casilla
		init();

		this.tipo = tipo;
		this.tablero = tablero;
		this.texto = texto;
		this.valor = valor;
  }
  
  Sorpresa(TipoSorpresa tipo, Tablero tablero){ //Envia carcel
		this(tipo, tablero, -1, null);
  }
  
  Sorpresa(TipoSorpresa tipo, int valor, String texto){ //Resto de sorpresas
		this(tipo, null, valor, texto);
  }
  
  Sorpresa(TipoSorpresa tipo, MazoSorpresas mazo){ //Evitar carcel
		this(tipo, null, -1, null);
		this.mazo = mazo;
  }
  
  public boolean jugadorCorrecto(int actual, ArrayList<Jugador> todos){
		return (actual >= 0 && actual < todos.size());
  }
  
  private void informe(int actual, ArrayList<Jugador> todos){
		Diario.getInstance().ocurreEvento(todos.get(actual).getNombre() + " ha activado una sorpresa. ");
  }
  
  void aplicarAJugador(int actual, ArrayList<Jugador> todos){
		switch(tipo){         
			case IRCARCEL:
				aplicarAJugador_irCarcel(actual,todos);
            
			case IRCASILLA:
        aplicarAJugador_irACasilla(actual, todos);
            
			case PAGARCOBRAR:
        aplicarAJugador_pagarCobrar(actual, todos);
            
			case PORCASAHOTEL:
        aplicarAJugador_porCasaHotel(actual, todos);
            
			case PORJUGADOR:
        aplicarAJugador_porJugador(actual, todos);
		
			case SALIRCARCEL:
        aplicarAJugador_salirCarcel(actual, todos);	
		}	
  }
  
  private void aplicarAJugador_irCarcel(int actual, ArrayList<Jugador> todos){
		if(jugadorCorrecto(actual, todos)){
			informe(actual, todos);
			todos.get(actual).encarcelar(tablero.getCarcel());
		}
  }
  
  private void aplicarAJugador_irACasilla(int actual, ArrayList<Jugador> todos){
		if(jugadorCorrecto(actual, todos)){
			informe(actual, todos);
	  
			int casillaActual = todos.get(actual).getNumCasillaActual();
			int tirada = tablero.calcularTirada(casillaActual, valor);
			int nuevaPos = tablero.nuevaPosicion(casillaActual, tirada);
	  
			todos.get(actual).moverACasilla(nuevaPos);
	  
			tablero.getCasilla(nuevaPos).recibeJugador(actual, todos);
		}
  }
  
  private void aplicarAJugador_pagarCobrar(int actual, ArrayList<Jugador> todos){
		if(jugadorCorrecto(actual, todos)){
			informe(actual, todos);
			todos.get(actual).modificarSaldo(valor);
		}
  }
  
  private void aplicarAJugador_porCasaHotel(int actual, ArrayList<Jugador> todos){
		if(jugadorCorrecto(actual, todos)){
			informe(actual, todos);
			todos.get(actual).modificarSaldo(valor * todos.get(actual).cantidadCasasHoteles());
		}
  }
  
  private void aplicarAJugador_porJugador(int actual, ArrayList<Jugador> todos){
		if(jugadorCorrecto(actual, todos)){
			informe(actual, todos);
		
			Sorpresa pagar = new Sorpresa(TipoSorpresa.PAGARCOBRAR, -1 * valor, "PAGAR");
			Sorpresa cobrar = new Sorpresa(TipoSorpresa.PAGARCOBRAR, (todos.size() - 1) * valor, "COBRAR");
		
			for(int i = 0; i < todos.size(); i++)
				if(todos.get(i) != todos.get(actual))
					pagar.aplicarAJugador_pagarCobrar(i, todos);
		
			cobrar.aplicarAJugador_pagarCobrar(actual, todos);
		}
  }
  
  private void aplicarAJugador_salirCarcel(int actual, ArrayList<Jugador> todos){
		if(jugadorCorrecto(actual, todos)){
			informe(actual, todos);
	  
			boolean haySalvoConducto = false;
	  
			for(int i = 0; i < todos.size() && !haySalvoConducto; i++)
				if(todos.get(i).tieneSalvoconducto())
					haySalvoConducto = true;
	  
			if(!haySalvoConducto){
				todos.get(actual).obtenerSalvoconducto(this);
				salirDelMazo();
			}
		}
  }
  
  void salirDelMazo(){
		if(tipo == TipoSorpresa.SALIRCARCEL)
			mazo.inhabilitarCartaEspecial(this);
  }
  
  @Override
  public String toString(){
		return "Sorpresa: " + tipo + "\n"; 
  }
  
  void usada(){
		if(tipo == TipoSorpresa.SALIRCARCEL)
			mazo.habilitarCartaEspecial(this);
  }
  
 
}
